src/libbson/src/bson/bson.c

```diff
@@ -60,6 +61,36 @@ typedef struct {
  */
 static const uint8_t gZero;
 
+/*
+ *--------------------------------------------------------------------------
+ *
+ * _bson_round_up_alloc_size --
+ *
+ *       Given a potential allocation length in bytes, round up to the
+ *       next power of two without exceeding BSON_MAX_SIZE.
+ *
+ * Returns:
+ *       If the input is <= BSON_MAX_SIZE, returns a value >= the input
+ *       and still <= BSON_MAX_SIZE. If the input was greater than
+ *       BSON_MAX_SIZE, it is returned unmodified.
+ *
+ * Side effects:
+ *       None.
+ *
+ *--------------------------------------------------------------------------
+ */
+
+static BSON_INLINE size_t
+_bson_round_up_alloc_size (size_t size)
+{
+   if (size <= BSON_MAX_SIZE) {
```

Comment:
Given the precondition on `bson_next_power_of_two`:

> It is up to the caller to guarantee this will not overflow.

suggest adding a static assertion to document and assert the call to `bson_next_power_of_two` with `BSON_MAX_SIZE` is always valid, e.g.:

```c
// BSON_MAX_SIZE ^ 2 must be representable for computational purposes even if not a valid BSON size.
BSON_STATIC_ASSERT2 (bson_max_size_is_doubleable, (SIZE_MAX / 2u) >= BSON_MAX_SIZE);
```

Aside: the "32-bit unsigned integer of required bytes" description of the argument to `bson_next_power_of_two` seems to be outdated.

Reply:
I do not think `mcommon_next_power_of_two_u32` is equivalent, as `bson_next_power_of_two ` currently supports `sizeof(size_t) == sizeof(uint64_t)` with the `BSON_WORD_SIZE` special-case in its definition. This assertion is to statically assert (rather than runtime checks) the precondition for the call to `bson_next_power_of_two` which does not perform overflow safety checks.

Reply:
> Do we intend to provide a way to increase BSON_MAX_SIZE?

Likely no. That would be a major API breaking change. The static assertion (or any similar runtime checks on `BSON_MAX_SIZE`) is just to both document and protect against any such future modifications to `BSON_MAX_SIZE`. I believe it is fine to defer changes that extend forward compatibility for related functions to support larger `BSON_MAX_SIZE` than what it is currently set to (which is sufficiently small enough to be handled even by 32-bit `size_t`).